src/EditorFeatures/Core/FindUsages/FindUsagesHelpers.cs

```diff
@@ -47,95 +47,111 @@ public static string GetDisplayName(ISymbol symbol)
+        public static async Task<(Solution solution, SymbolAndProjectId symboAndProjectId, ImmutableArray<SymbolAndProjectId> implementations, string message)?> FindSourceImplementationsAsync(Document document, int position, CancellationToken cancellationToken)
         {
-            var symbolAndProject = await GetRelevantSymbolAndProjectAtPositionAsync(
+            var symbolAndProjectOpt = await GetRelevantSymbolAndProjectAtPositionAsync(
                 document, position, cancellationToken).ConfigureAwait(false);
-            if (symbolAndProject == null)
-            {
+            if (symbolAndProjectOpt == null)
                 return null;
-            }
 
-            return await FindImplementationsAsync(
-                symbolAndProject?.symbol, symbolAndProject?.project, cancellationToken).ConfigureAwait(false);
+            var (symbol, project) = symbolAndProjectOpt.Value;
+            var symbolAndProjectId = new SymbolAndProjectId(symbol, project.Id);
+            return await FindSourceImplementationsAsync(
+                symbolAndProjectId, project.Solution, cancellationToken).ConfigureAwait(false);
         }
 
-        private static async Task<(ISymbol symbol, Project project, ImmutableArray<ISymbol> implementations, string message)?> FindImplementationsAsync(
-            ISymbol symbol, Project project, CancellationToken cancellationToken)
+        private static async Task<(Solution solution, SymbolAndProjectId symbolAndProjectId, ImmutableArray<SymbolAndProjectId> implementations, string message)?> FindSourceImplementationsAsync(
+            SymbolAndProjectId symbolAndProjectId, Solution solution, CancellationToken cancellationToken)
         {
-            var implementations = await FindImplementationsWorkerAsync(
-                symbol, project, cancellationToken).ConfigureAwait(false);
+            var builder = new HashSet<SymbolAndProjectId>(SymbolAndProjectIdComparer.SymbolEquivalenceInstance);
 
-            var filteredSymbols = implementations.WhereAsArray(
-                s => s.Locations.Any(l => l.IsInSource));
+            // Find the direct implementations first.
+            builder.AddRange(await FindSourceImplementationsWorkerAsync(
+                symbolAndProjectId, solution, cancellationToken).ConfigureAwait(false));
 
-            return filteredSymbols.Length == 0
-                ? (symbol, project, filteredSymbols, EditorFeaturesResources.The_symbol_has_no_implementations)
-                : (symbol, project, filteredSymbols, null);
+            // If we're in a linked file, try to find all the symbols this links to, and find all the implementations of
+            // each of those linked symbols. De-dupe the results so the user only gets unique results.
+            var linkedSymbols = await SymbolFinder.FindLinkedSymbolsAsync(
+                symbolAndProjectId.Symbol, solution, cancellationToken).ConfigureAwait(false);
```

Comment:
❔ So `FindLinkedSymbolsAsync` never returns the input symbol? Seems like this code would be simpler if it could just start with getting `linkedSymbols` and then iterate over it.

Reply:
i acutally went back and forth on that, but ended up here... neither felt super satisfying...